src/plugins/ssh/index.ts

```diff
@@ -203,11 +145,60 @@ async function spawnSshNode(
     );
 
     // TODO ENG-2284 support login with Google Cloud: currently return a boolean to indicate if the exception was a Google login error.
-    const { isAccessPropagated, isGoogleLoginException } =
-      accessPropagationGuard(child, options.debug);
+    let isEphemeralAccessDeniedException = false;
+    let isGoogleLoginException = false;
+
+    let timeout: NodeJS.Timeout;
+
+    const remoteCommandOption = hasRemoteCommandOption(options.args);
+
+    const resetTimeout = () => {
+      if (!remoteCommandOption) {
+        return;
+      }
+      if (timeout) {
+        clearTimeout(timeout);
+      }
+      timeout = setTimeout(() => {
+        print2("SSH connection established.");
+      }, STDERR_TIMEOUT_PERIOD);
+    };
+
+    const beforeStart = Date.now();
+    child.stderr.on("data", (chunk) => {
+      resetTimeout();
+
+      const chunkString: string = chunk.toString("utf-8");
+
+      if (options.debug) print2(chunkString);
+
+      const match = UNPROVISIONED_ACCESS_MESSAGES.find((message) =>
+        chunkString.match(message.pattern)
+      );
+
+      if (
+        match &&
+        Date.now() <=
+          beforeStart +
+            (match.validationWindowMs || DEFAULT_VALIDATION_WINDOW_MS)
+      ) {
+        isEphemeralAccessDeniedException = true;
+      }
+
+      const googleLoginMatch = chunkString.match(GOOGLE_LOGIN_MESSAGE);
+      isGoogleLoginException = isGoogleLoginException || !!googleLoginMatch; // once true, always true
+      if (isGoogleLoginException) {
+        // always overwrite to false so we don't retry the access
+        isEphemeralAccessDeniedException = false;
+      }
```

Comment:
Can we keep this GCP-specific logic within the GCP-specific SSH provider? We should ensure separation of concerns, and keep anything CSP-specific within that provider's plugin.

I know this is just a code move, but we should clean up as we go along.